src/differential_arithmetic.jl

```diff
@@ -65,12 +66,13 @@ for T in (:One, :AbstractThunk, :Composite, :Any)
     @eval Base.:-(::Zero, b::$T) = -b
     @eval Base.:-(a::$T, ::Zero) = a
 
-    @eval Base.:*(::Zero, ::$T) = Zero()
-    @eval Base.:*(::$T, ::Zero) = Zero()
-
     @eval LinearAlgebra.dot(::Zero, ::$T) = Zero()
     @eval LinearAlgebra.dot(::$T, ::Zero) = Zero()
 end
+for T in (:One, :AbstractThunk, :Composite, :Number)
```

Comment:
I don't think this is ok.
At very least we need to accept `AbstractMatrix` I think.
And semantically we *should* accept `Any`.